### PR TITLE
Use Google Maps API Autocomplete and Geocoder

### PR DESCRIPTION
**Pull request by ianguerin**
Include additional Google Maps API libraries on page.
Create model services to interact with Google Maps APIs.

NCEAS#1796

### CODE REVIEW

File: src/js/models/maps/viewfinder/ViewfinderModel.js
Comment (yvonnesjy) [inline src/js/models/maps/viewfinder/ViewfinderModel.js:24]:
```suggestion
       * in the list of predictions that should be highlighted as focus.
```

**Review by yvonnesjy — CHANGES_REQUESTED** [review]

File: src/js/models/geocoder/GeocodedLocation.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocodedLocation.js:23]: The object is used to instantiate a GeoBoundingBox, but is not itself a GeoBoundingBox

File: src/js/models/geocoder/GeocodedLocation.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocodedLocation.js:14]: Adding JSdocs for all defaults functions

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:26]: I haven't seen any places where plan classes are used, though I do agree, this probably warrants more discussion.. @robyngit for any thoughts you may have

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:53]: Done

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:69]: Can do

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:81]: Done

File: src/js/models/maps/viewfinder/ViewfinderModel.js
Comment (ianguerin) [reply src/js/models/maps/viewfinder/ViewfinderModel.js:24]: Done

File: src/js/models/geocoder/GeocoderSearch.js
Comment (yvonnesjy) [reply src/js/models/geocoder/GeocoderSearch.js:23]: hmmm yea- if not `class`, prototypes without Backbone Model exists, ex. Utilities.js. Though i don't see why we can't just use a class.

Comment (robyngit) [reply src/js/models/geocoder/GeocoderSearch.js:23]: Could you see us using some of the backbone model functionality as we add different geocoding services in the future, for example, listening to when a new geocoder model is added as an attribute to this model?

File: src/js/models/geocoder/GeocoderSearch.js
Comment (robyngit) [inline src/js/models/geocoder/GeocoderSearch.js:15]: `Prediction` is missing a dependency in the array and isn't used here (except in JSdocs)

Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:15]: Done

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:23]: I don't think it will ever need any of the Model capabilities. I'll migrate to a JS `class`, where should it live though? If it's not a model, I think I should move it out of js/models, is js/geocoder fine to create? Or should I put it next to Utilities.js in js/common?

File: src/js/models/geocoder/GoogleMapsAutocompleter.js
Comment (yvonnesjy) [inline src/js/models/geocoder/GoogleMapsAutocompleter.js:12]: Care to update this and Geocoder to use ES6 class as well?

File: src/js/models/geocoder/GeocoderSearch.js
Comment (yvonnesjy) [reply src/js/models/geocoder/GeocoderSearch.js]: Hm I see this resolved but no commit yet. Forgot to push?

**Review by yvonnesjy — CHANGES_REQUESTED** [review]

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js]: It's there, https://github.com/NCEAS/metacatui/pull/2291/commits/7dacf8eb4ae4509cfbbcb06b53137c6e192cbcde

File: src/js/models/geocoder/GoogleMapsAutocompleter.js
Comment (ianguerin) [reply src/js/models/geocoder/GoogleMapsAutocompleter.js]: Done

File: src/js/models/geocoder/GeocoderSearch.js
Comment (yvonnesjy) [reply src/js/models/geocoder/GeocoderSearch.js]: oh okay I see the disconnect- I think of GeocoderSearch as the layer of abstraction from the autocomplete and geocode service, making future changes easier if we ever want to move away from GMap. so I think here it probably shouldn't access prediction.placeId, since it's a GMap only concept? Can it just pass the prediction along and let the google service wrapper do the unpacking?

Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js]: I agree that GeocoderSearch is the abstraction layer, anywhere GeocoderSearch is used doesn't need to know about Google Maps. GeocoderSearch _does_ need to know about Google Maps though, so having it access the Places ID is okay. In any case where we change the geocoder backend later, GeocoderSearch will need to be updated. I do think your request to pass in a Prediction here helps in that case, since the call site doesn't need to be updated if we ever change geocoder backend.

Comment (yvonnesjy) [reply src/js/models/geocoder/GeocoderSearch.js]: hmm I think knowing which service to use (google maps), and knowing the service implementation (.placeId) are different. imo, the sole purpose of GeocoderSearch is to decide which service to use. ideally, we just swap out new GoogleMapsGeocoder() for new BingMapsGeocoder() and expect it work OOTB.

that said, I'm only arguing for the principles at this point- I know the difference is very very minor. I'll leave it up to you since we've agreed that GeocoderSearch is the abstraction layer.

Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js]: Thanks for the explanation, updated

**Review by yvonnesjy — APPROVED** [review]

**Review by robyngit — CHANGES_REQUESTED** [review]
Beautiful, @ianguerin! Appreciate your consistently thorough unit tests! Also really nice to have working examples of how we can start using es6 classes where appropriate. Other than the very in-line comments, one request I have for you is to ensure that the new `@class`es are showing up in docs site as expected. Currently the new backbone models are sorted under models/geocoder as expected:

Screenshot 2024-03-13 at 11 45 20

But the es6 classes are thrown at the end of the list (appear as though under "views"):

Screenshot 2024-03-13 at 11 45 58

and also none of the documented properties are visible:

Screenshot 2024-03-13 at 11 58 09

Instructions to build docs locally are in [`docs/README.md`](https://github.com/NCEAS/metacatui/tree/main/docs#readme).

Perhaps these points are already taken into account for the subsequent PRs, but I just wanted to mention now in case:

- We need to handle the situation where a repository does not have a google maps API key configured
- Since autocomplete functions could generate a high volume of API requests, maybe we'll need to do some debouncing to limit the number of requests made as the user types.

**Comment by ianguerin** [comment]
Well, I've figured out the documentation issues. I had to declare the class and return it on separate lines in the require js module. That and I was missing @classcategory declarations on a few files

File: src/js/models/geocoder/GoogleMapsAutocompleter.js
Comment (ianguerin) [reply src/js/models/geocoder/GoogleMapsAutocompleter.js:32]: Other than logging to the console (which I'd prefer not to do in case it gets logged in production), is there a place to log errors? Otherwise I'm unsure of the benefit there. I'd prefer to not show the user anything if its not an error that they can recover from. As far as I can tell from looking at the [Google Maps docs](https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompleteResponse), this will just return an empty response if the input is bad. So the only failure here would be a network failure

File: src/js/models/geocoder/GeocoderSearch.js
Comment (ianguerin) [reply src/js/models/geocoder/GeocoderSearch.js:1]: In case where you want to switch from Google Maps autocomplete or geocoder you'd need to change the implementation of GeocoderSearch, but the idea is that any code that relies on GeocoderSearch does not need to change, they'd still call GeocoderSearch#autocomplete, for example. Happy to write a guide though. If possible I'd like to do that in a subsequent PR so that I can spend time writing it that does not block this PR

**Comment by ianguerin** [comment]
> Perhaps these points are already taken into account for the subsequent PRs, but I just wanted to mention now in case:
> 
> * We need to handle the situation where a repository does not have a google maps API key configured
> * Since autocomplete functions could generate a high volume of API requests, maybe we'll need to do some debouncing to limit the number of requests made as the user types.

1- I'd imagine that the repo that doesn't have Google Maps API key configured would also not turn on the viewfinder feature. 

2- Totally agreed, in a later PR I denounce the call to autocomplete by 250ms to prevent sending a request on each keystroke.

**Pull request merged**
